src/constitutive/TACSIsoShellConstitutive.cpp

```diff
@@ -321,40 +329,49 @@ TacsScalar TACSIsoShellConstitutive::evalFailureStrainSens( int elemIndex,
     TacsScalar top = properties->vonMisesFailure2D(st);
     TacsScalar bottom = properties->vonMisesFailure2D(sb);
 
+    TacsScalar ksMax;
     if (TacsRealPart(top) > TacsRealPart(bottom)){
-      TacsScalar psi[3], phi[3];
-      properties->vonMisesFailure2DStressSens(st, psi);
-      mat3x3SymmMult(C, psi, phi);
+      ksMax = top;
+    }
+    else {
+      ksMax = bottom;
+    }
 
-      sens[0] = phi[0];
-      sens[1] = phi[1];
-      sens[2] = phi[2];
+    // Use a ks approximation for the max value
+    TacsScalar ksSum = exp(ksWeight*(top - ksMax)) + exp(ksWeight*(bottom - ksMax));
+    TacsScalar ksVal = ksMax + log(ksSum) / ksWeight;
 
-      sens[3] = ht*phi[0];
-      sens[4] = ht*phi[1];
-      sens[5] = ht*phi[2];
+    TacsScalar psi[3], phi[3];
 
-      sens[6] = sens[7] = sens[8] = 0.0;
+    // Contribution from plate top
+    properties->vonMisesFailure2DStressSens(st, psi);
+    mat3x3SymmMult(C, psi, phi);
+    TacsScalar ksFactor = exp(ksWeight*(top - ksMax)) / ksSum;
 
-      return top;
-    }
-    else {
-      TacsScalar psi[3], phi[3];
-      properties->vonMisesFailure2DStressSens(sb, psi);
-      mat3x3SymmMult(C, psi, phi);
+    sens[0] = ksFactor*phi[0];
+    sens[1] = ksFactor*phi[1];
+    sens[2] = ksFactor*phi[2];
 
-      sens[0] = phi[0];
-      sens[1] = phi[1];
-      sens[2] = phi[2];
+    sens[3] = ksFactor*ht*phi[0];
+    sens[4] = ksFactor*ht*phi[1];
+    sens[5] = ksFactor*ht*phi[2];
 
-      sens[3] = -ht*phi[0];
-      sens[4] = -ht*phi[1];
-      sens[5] = -ht*phi[2];
+    // Contribution from plate bottom
+    properties->vonMisesFailure2DStressSens(sb, psi);
+    mat3x3SymmMult(C, psi, phi);
+    ksFactor = exp(ksWeight*(bottom - ksMax)) / ksSum;
 
-      sens[6] = sens[7] = sens[8] = 0.0;
+    sens[0] += ksFactor*phi[0];
+    sens[1] += ksFactor*phi[1];
+    sens[2] += ksFactor*phi[2];
 
-      return bottom;
-    }
+    sens[3] += ksFactor*-ht*phi[0];
```

Comment:
the *-ht* is a bit confusing here.. can we just make it
sens[3] -= ksFactor*ht*psi[0];

Reply:
Should be addressed in last commit